src/Microsoft.FileFormats/PDB/MSFZFile.cs

```diff
@@ -0,0 +1,408 @@
+            // Find the fragments for this stream.
+            uint fragmentIndex = _streamDirStarts[stream];
+            uint totalBytesTransferred = 0;
+
+            while (count != 0)
+            {
+                uint fragmentSize = _streamDir[fragmentIndex];
+                if (fragmentSize == MsfzConstants.NilFragmentSize || fragmentSize == 0)
+                {
+                    break;
+                }
+
+                // We can safely index into _streamDir because we have already validated its contents,
+                // when we first opened the MSFZ file.
+
+                MsfzFragmentLocation fragmentLocation;
+                fragmentLocation.Low = _streamDir[fragmentIndex + 1];
+                fragmentLocation.High = _streamDir[fragmentIndex + 2];
+                fragmentIndex += MsfzConstants.UInt32PerFragmentRecord;
+
+                // If the position is beyond the range of bytes covered by this fragment, then skip
+                // this fragment and adjust our read position.
+                if (position >= fragmentSize)
+                {
+                    position -= fragmentSize;
+                    continue;
+                }
+
+                // We found the fragment that we are going to read from.
+                uint transferSize = Math.Min(count, (uint)(fragmentSize - position));
+
+                // Is this fragment compressed or uncompressed?
+                if (fragmentLocation.IsCompressedChunk)
+                {
+                    // Compressed fragments are not supported by this implementation.
+                    throw new NotSupportedException($"This implementation does not support decompression of PDZ (compressed PDB) files. "
+                        + $"It can only read data from uncompressed streams within PDZ files. (stream {stream}, position {position})");
+                }
+
+                ulong fileOffset = fragmentLocation.UncompressedFileOffset + position;
+
+                uint fileBytesTransferred = _reader.Read(fileOffset, buffer, bufferOffset, transferSize);
+                if (fileBytesTransferred != transferSize)
+                {
+                    // We expect the entire read to be satisfied.
+                    throw new Exception("Internal error in MSFZ reader. The underlying file reader did not read enough data from the file.");
+                }
+
+                count -= transferSize;
+                bufferOffset += transferSize;
+                position += transferSize;
+                totalBytesTransferred += transferSize;
+            }
+
+            return totalBytesTransferred;
+        }
+
+
+        public void Dispose()
+        {
+            if (_reader.DataSource is IDisposable disposable)
+            {
+                disposable.Dispose();
+            }
+        }
+    }
+
+    /// <summary>
+    /// Describes the MSFZ File Header on-disk data structure.
+    /// </summary>
+    public sealed class MSFZFileHeader : TStruct
```

Comment:
```suggestion
    internal sealed class MSFZFileHeader : TStruct
```
This looked like new public API - if we need to do that it would be good to understand what is consuming the API.